File: day2_part1_2018.dart

import 'dart:io';

void main() {
  List<String> boxIds = File('input.txt').readAsLinesSync();
  
  int twoCount = 0;
  int threeCount = 0;
  
  for (String id in boxIds) {
    Map<String, int> charCount = {};
    
    for (int i = 0; i < id.length; i++) {
      charCount.update(id[i], (value) => value + 1, ifAbsent: () => 1);
    }
    
    if (charCount.containsValue(2)) {
      twoCount++;
    }
    
    if (charCount.containsValue(3)) {
      threeCount++;
    }
  }
  
  int checksum = twoCount * threeCount;
  print(checksum);
}

File: objectivec/day2_part1_2018.m
#import <Foundation/Foundation.h>

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSString *input = [NSString stringWithContentsOfFile:@"input.txt" encoding:NSUTF8StringEncoding error:nil];
        NSArray *boxIDs = [input componentsSeparatedByString:@"\n"];
        
        int twoCount = 0;
        int threeCount = 0;
        
        for (NSString *boxID in boxIDs) {
            NSMutableDictionary *letterCounts = [NSMutableDictionary dictionary];
            for (int i = 0; i < boxID.length; i++) {
                NSString *letter = [boxID substringWithRange:NSMakeRange(i, 1)];
                if (letterCounts[letter]) {
                    letterCounts[letter] = @([letterCounts[letter] intValue] + 1);
                } else {
                    letterCounts[letter] = @1;
                }
            }
            
            BOOL hasTwo = NO;
            BOOL hasThree = NO;
            for (NSNumber *count in [letterCounts allValues]) {
                if ([count intValue] == 2) {
                    hasTwo = YES;
                } else if ([count intValue] == 3) {
                    hasThree = YES;
                }
            }
            
            if (hasTwo) {
                twoCount++;
            }
            if (hasThree) {
                threeCount++;
            }
        }
        
        printf("Checksum: %d\n", twoCount * threeCount);
    }
    return 0;
}